Created flows for creating and deleting tags

annadoesdesign opened the pull request:
- Create Tag Button
- Delete Menu Item Added
- Create Tag Modal created with name, description, color, owners, and entities 
-
<img width="1840" alt="Screenshot 2025-04-07 at 12 34 00 PM" src="https://github.com/user-attachments/assets/7e1d79b7-29b6-4648-aeb8-89b4e87e1aaa" />
<img width="1840" alt="Screenshot 2025-04-07 at 12 34 14 PM" src="https://github.com/user-attachments/assets/91cb6feb-f01b-4f5c-b476-093a148345a8" />
<img width="1840" alt="Screenshot 2025-04-07 at 12 34 09 PM" src="https://github.com/user-attachments/assets/dd390a68-01ec-4d48-bec3-8c106e51c9cd" />
<img width="1840" alt="Screenshot 2025-04-07 at 12 34 03 PM" src="https://github.com/user-attachments/assets/9cdb029a-b45d-4c45-b7d2-7763107f635d" />


- [ ] The PR conforms to DataHub's [Contributing Guideline](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md) (particularly [Commit Message Format](https://github.com/datahub-project/datahub/blob/master/docs/CONTRIBUTING.md#commit-message-format))
- [ ] Links to related issues (if applicable)
- [ ] Tests for the changes have been added/updated (if applicable)
- [ ] Docs related to the changes have been added/updated (if applicable). If a new feature has been added a Usage Guide has been added for the same.
- [ ] For any breaking change/potential downtime/deprecation/big changes an entry has been made in [Updating DataHub](https://github.com/datahub-project/datahub/blob/master/docs/how/updating-datahub.md)

jjoyce0510 commented on datahub-web-react/src/app/tags/CreateNewTagModal/TagDetailsSection.tsx:
```diff
@@ -0,0 +1,74 @@
+        <SectionContainer>
+            <FormSection>
+                <Input
+                    label="Name *"
```
Why name *?

Not just "Name"

jjoyce0510 commented on datahub-web-react/src/app/tags/CreateNewTagModal/index.tsx:
```diff
@@ -0,0 +1,159 @@
+
+    // Check permissions using UserContext
+    const userContext = useUserContext();
+    const canCreateTags = userContext?.platformPrivileges?.createTags || userContext?.platformPrivileges?.manageTags;
```
nice!!!

jjoyce0510 commented on datahub-web-react/src/app/tags/CreateNewTagModal/index.tsx:
```diff
@@ -0,0 +1,159 @@
+     * Handler for creating the tag and applying it to entities
+     */
+    const onOk = async () => {
+        if (!canCreateTags) {
```
Why not handle this before teh user even hits the create button? 
 Maybe the create button is fully disabled, unclickable if the user does not have privileges?

And in such case, we can also add a Tooltip that tells the user: "You do not have permission to create tags."

jjoyce0510 commented on datahub-web-react/src/app/tags/CreateNewTagModal/index.tsx:
```diff
@@ -0,0 +1,159 @@
+            const newTagUrn = createTagResult.data?.createTag;
+
+            if (!newTagUrn) {
+                throw new Error('Failed to create tag: No URN returned');
```
Don't throw this error. 

Just message.error("Failed to create tag. An unexpected error occurred").

Try to never throw exceptions in the code, it will cause the user experience to completely bomb out.

jjoyce0510 commented on datahub-web-react/src/app/tags/CreateNewTagModal/index.tsx:
```diff
@@ -0,0 +1,159 @@
+    if (!canCreateTags) {
+        return (
+            <Modal title="Create New Tag" onCancel={onClose} buttons={buttons} open={open} centered width={400}>
+                <p>You don&apos;t have permission to create tags. Please contact your DataHub administrator.</p>
```
You should do this stuff before the user even gets into the modal ideally, in the parent ManageTags component!

jjoyce0510 commented on datahub-web-react/src/app/tags/ManageTags.tsx:
```diff
@@ -55,6 +57,11 @@ const ManageTags = () => {
+
+    // Check permissions using UserContext
+    const userContext = useUserContext();
+    const canCreateTags = userContext?.platformPrivileges?.createTags || userContext?.platformPrivileges?.manageTags;
```
oh nice!! you also have this here.

jjoyce0510 commented on datahub-web-react/src/app/tags/ManageTags.tsx:
```diff
@@ -148,6 +193,14 @@ const ManageTags = () => {
+                open={showCreateTagModal}
+                onClose={() => {
+                    setShowCreateTagModal(false);
+                    refetch(); // Refresh the tag list after creating a new tag
```
tip: You might want to wait a few seconds before refetching. That's because we have some delay on the backend after creating a tag. So if you immediately refetch, the new tag may not have been fully added and synced in the backend.

You can do something like: 

setTimeout(() => refetch(), 3000).

Ideally, we'd use the Apollo cache here and simply insert a new tag into the cached apollo query response. Claude may be able to help here - not sure!

jjoyce0510 commented on datahub-web-react/src/app/tags/TagsTable.tsx:
```diff
@@ -62,6 +78,42 @@ const TagsTable = ({ searchQuery, searchData, loading: propLoading, networkStatu
+    // Function to delete a tag with confirmation
+    const deleteTag = useCallback(
+        (tagUrn: string) => {
+            const fullDisplayName = entityRegistry.getDisplayName(EntityType.Tag, { urn: tagUrn } as any);
```
const fullDisplayName = entityRegistry.getDisplayName(EntityType.Tag, { urn: tagUrn } as any);

don't do this - we should pass in the actual tag entity here.

const fullDisplayName = entityRegistry.getDisplayName(EntityType.Tag, tag);

You can get the tag entity from tagsData. Inside one object in there you'll have an entity field.

tagsData.map(tagData => tagData.entity // This is the object you want to pass in)

jjoyce0510 commented on datahub-web-react/src/app/tags/TagsTable.tsx:
```diff
@@ -62,6 +78,42 @@ const TagsTable = ({ searchQuery, searchData, loading: propLoading, networkStatu
+
+            // Extract just the tag name without the URN prefix
+            // This assumes URNs are in the format "urn:li:tag:TagName"
+            const displayName = tagUrn.startsWith('urn:li:tag:') ? tagUrn.replace('urn:li:tag:', '') : fullDisplayName;
```
Why this? Just always use full display name!

jjoyce0510 commented on datahub-web-react/src/app/tags/TagsTable.tsx:
```diff
@@ -119,12 +171,19 @@ const TagsTable = ({ searchQuery, searchData, loading: propLoading, networkStatu
                             }}
+                            onDelete={() => {
+                                if (canManageTags) {
+                                    deleteTag(record.entity.urn);
```
Can we make it such that the delete button is disabled altogether if the user does not have permission to delete?

jjoyce0510 replied:
I see - we need this for the modal display. Hmm maybe there is no other way that's easy

jjoyce0510 commented on datahub-web-react/src/app/tags/TagsTable.tsx:
```diff
@@ -154,6 +213,35 @@ const TagsTable = ({ searchQuery, searchData, loading: propLoading, networkStatu
+            {/* Delete confirmation modal */}
+            {showDeleteModal && tagToDelete && (
+                <Modal
+                    title={`Delete tag: ${tagToDelete.displayName}`}
```
Copy: "Delete tag {Tag display name}". No colon needed.

jjoyce0510 commented on datahub-web-react/src/app/tags/TagsTableColumns.tsx:
```diff
@@ -225,26 +225,36 @@ export const TagAppliedToColumn = React.memo(({ tagUrn }: { tagUrn: string }) =>
+            {
+                key: '1',
+                label: (
+                    <MenuItem onClick={onDelete} data-testid="action-delete" style={{ color: colors.red[500] }}>
```
here's where we should say disabled={!canDeleteTag} or whatever

jjoyce0510 reviewed (commented):
Overall, looking good. Few serious changes.

We'll also need to watch out for the smoke test for managing tags. If that doesn't work, we'll need to update it!

jjoyce0510 commented on datahub-web-react/src/app/tags/CreateNewTagModal/TagDetailsSection.tsx:
```diff
@@ -0,0 +1,74 @@
+/**
+ * Component for tag name, description, and color selection
+ */
+const TagDetailsSection: React.FC<TagDetailsProps> = ({
```
I'd prefer if we define the types TagDetailsProps etc inside the same file where the component is defined. That's the usual pattern. Currently it's defined inside that dedicated types.ts file!

jjoyce0510 commented on datahub-web-react/src/app/tags/TagsTable.tsx:
```diff
@@ -62,6 +78,45 @@ const TagsTable = ({ searchQuery, searchData, loading: propLoading, networkStatu
+
+            // Using the modal from @components
+            setShowDeleteModal(true);
+            setTagToDelete({
```
doing this whole tagToDelete state management feels a bit overly complicated. I wonder if there's a simpler approach Claude can think of

jjoyce0510 reviewed (commented):
Overall looking quite good. Just have a few comments on the specifics around implementation

annadoesdesign commented:
@jjoyce0510 here is the update: 
1. OwnersSection.tsx
Replaced separate search queries for users and groups with a single searchAcrossEntities query
Removed redundant state by using a single search results array instead of combining two
Updated the search function to pass both entity types (CorpUser and CorpGroup) in a single API call
Simplified loading state management
Made the component more efficient by reducing API calls
2. TagDetailsSection.tsx
Moved the TagDetailsProps interface from types.ts directly into the component file
Made the component more self-contained following the pattern suggested in PR feedback
Co-located the component definition with its type definition for easier maintenance
3. types.ts
Removed the TagDetailsProps interface (now defined in TagDetailsSection.tsx)
Kept other interface definitions that are still shared across components
4. TagsTable.tsx
Simplified tag deletion state management
Split tag deletion logic into two clear functions:
showDeleteConfirmation: Sets up the deletion and shows modal
handleDeleteTag: Performs the actual deletion
Replaced complex tagToDelete object state with two simple string states:
tagUrnToDelete: For storing just the URN
tagDisplayName: For UI display purposes
Improved modal handling by removing conditional rendering
Reduced dependency array size for callbacks, making the code more maintainable
Fixed the button handler to call a function directly rather than accessing it through state
